zne/utils/strategy.py

```diff
@@ -0,0 +1,278 @@
+    04. ``__repr__``: a string representation of the strategy instance.
+    05. ``__eq__``: to compare strategy instances.
+    06. ``NAME``: the name of the strategy (i.e. class name).
+    07. ``INIT_NAMESPACE``: a tuple of the strategy's settings names.
+    08. ``SETTINGS_NAMESPACE``: a tuple of the strategy's settings names.
+    09. ``settings``: a dictionary of settings and their corresponding values/states.
+    10. ``init_args``: a copy of the construction-time init args.
+    11. ``replicate``: to build replicas of any strategy instance with updated settings.
+
+Out of these, {01, 02, 03} extend those found in the decorated class, and {04, 05}
+can be overriden; all others should be left unmodified.
+
+Classes inheriting from a :class:`strategy` class, are strategy classes
+themselves as well; with settings either inherited or provided by a new
+``__init__`` method. This means that they do not need to be re-decorated.
+
+Generally speaking, strategies are mutable, meaning that their settings
+can be updated at runtime. The implicit assumption made by this decorator is
+that the current state of all settings listed in ``SETTINGS_NAMESPACE`` can be
+accessed through a ``getattr`` call. If any of these settings is not accessible
+that way the corresponding (construction-time) init arg will be retrieved instead,
+possibly leading to conflicts and runtime errors if mutated after instantiation.
+If certain settings cannot be accessed by design, the suggested solution is
+removing said settings from the settings namespace as explained before, or adding a
+dummy property that retrieves their state (i.e. getter).
+
+Oftentimes, developers will need their strategies to be immutable after instantiation.
+To this end, the :class:`~zne.utils.strategy.strategy` decorator class accepts and
+optional keyword argument ``frozen`` which, if set to ``True`` (default ``False``),
+will forbid all ``setattr`` calls, and automatically return deepcopies of mutable
+settings. In this scenario, settings will still be declared through the class'
+``__init__`` method signature, while the body could be left blank (i.e. ``pass``),
+hold the appropriate docstring, or used for validation among other things.
+
+Replicas of any strategy instance can be produced by calling their
+:method:`~zne.utils.strategy.strategy.replicate` method with updated settings
+as keyword arguments. This is specially useful when dealing with frozen
+(i.e. immutable) strategies.
+"""
+
+from __future__ import annotations
+
+from copy import deepcopy
+from inspect import signature
+from typing import Any
+
+from .classconstant import classconstant
+from .unset import UNSET
+
+
+################################################################################
+## CLASS DECORATOR
+################################################################################
+class strategy:  # pylint: disable=invalid-name
+    """Strategy class decorator."""
```

Comment:
A more descriptive docstring would be helpful here.

Reply:
Done @garrison !